HPF handles searches from unreachable source cells into cut off areas.

RoosterDragon opened the pull request:
If a path search is performed by the HierarchicalPathFinder when the source cell is unreachable location, a path is still allowed and starts from one of the cells adjacent to the location. If moving into one of these cells results in the actor moving into an isolated area that cannot reach the target this would previously crash as no abstract path could be found. Now we handle such locations by giving them a unreachable cost so the path search will not attempt to explore them.

Imagine a map split into two by a one tile wide line of impassable cliffs. If an actor is on top of these cliffs it is allowed to path because it can "jump off" the cliff and move into the cell immediately either side of it. However it is important which side it chooses to jump into, as one it has moved off the cliff it loses access to the other side. The previous error came about because the path might try and search on the side that couldn't reach the target location. Now we avoid that being considered.

----
Fixes #20475
Fixes #20508
Fixes #20507
Fixes #20528

Testcase: Applying this diff onto the repo from the bug and running the replay no longer crashes, and the replay ends cleanly instead.

Request applying to stable as this is a bugfix for the new pathfinding.

----

Second fix added for 20528, can be confirmed by running the replay on that bug with the fix applied.

HPF is aware of map projection changes.

An event is added to Map to indicate when the cell projection is changed. This is important as this can mean Map.Contains(CPos) could now return different results for the cell. The HierarchicalPathFinder is made aware of these changes so it can rebuild any out-of-date information. This fixes prevent a crash if a cell that was previously outside the map changes height and becomes inside the map. The local path search will explore the cell as it is inside the map - but if the HPF was unaware if had been updated, it will still consider the cell to be outside the map and unreachable, resulting in a crash.

RoosterDragon committed:
HPF handles searches from unreachable source cells into cut off areas.

If a path search is performed by the HierarchicalPathFinder when the source cell is unreachable location, a path is still allowed and starts from one of the cells adjacent to the location. If moving into one of these cells results in the actor moving into an isolated area that cannot reach the target this would previously crash as no abstract path could be found. Now we handle such locations by giving them a unreachable cost so the path search will not attempt to explore them.

Imagine a map split into two by a one tile wide line of impassable cliffs. If an actor is on top of these cliffs it is allowed to path because it can "jump off" the cliff and move into the cell immediately either side of it. However it is important which side it chooses to jump into, as one it has moved off the cliff it loses access to the other side. The previous error came about because the path might try and search on the side that couldn't reach the target location. Now we avoid that being considered.

PunkPun reviewed (commented):
Can confirm that this fixes the crash reported in SP, but I was unable to reproduce the crash in RA. I tried various setups and just couldn't get bleed to crash

PunkPun commented:
Does this fix only affect mods with custom terrain layers?

Mailaender commented:
Confirmed with https://cdn.discordapp.com/attachments/279290580201373696/1049013821022416986/ra-2022-12-04T163540Z.orarep

PunkPun commented:
Fix confirmed from that replay. Crash happened from this cell and I assume this attack move order.

<img width="761" alt="Screenshot 2022-12-04 at 20 00 14" src="https://user-images.githubusercontent.com/37534529/205507565-2489b544-d0c5-4413-86fa-6fee3da3d1b2.png">

dnqbob commented:
> Does this fix only affect mods with custom terrain layers?

I think not. In my debug reprot:

```
The abstract path should never be searched for an unreachable point. Abstract cell 89,-49 failed to route to abstract cell. 
harv: 88,-50
```

Which means it is nothing related to custom terrain layers (SP also don't have custom terrain layer in engine), and related to common unit that used in all mods.

RoosterDragon commented:
> Can confirm that this fixes the crash reported in SP, but I was unable to reproduce the crash in RA. I tried various setups and just couldn't get bleed to crash

Testcase based on the description: Create a map with some units pre-placed on impassable terrain that divides the map into two. Try moving one of those units at least 20 cells so HPF kicks in. On bleed the crash should reproduce, on the units will path successfully.

![image](https://user-images.githubusercontent.com/3399086/205711831-00b6d013-f82f-40b4-a666-d545cce91583.png)

PunkPun commented on OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs:
```diff
@@ -1112,13 +1135,13 @@ List<GraphConnection> AbstractEdge(CPos abstractCell)
 						{
 							var adjacentSource = cell + dir;
-							if (!world.Map.Contains(adjacentSource))
+							if (!world.Map.Contains(adjacentSource) || unpathableNodes.Contains(adjacentSource))
```
this can crash with `System.NullReferenceException`

RoosterDragon replied:
If we pass non-null `sources`, we also provide non-null `unpathableNodes`.

PunkPun replied:
```suggestion
							if (!world.Map.Contains(adjacentSource) || (unpathableNodes != null && unpathableNodes.Contains(adjacentSource)))
```

PunkPun replied:
> If we pass non-null `sources`, we also provide non-null `unpathableNodes`.

Well the game just crashed for me

PunkPun replied:
And this suggested fix repaired my replay

RoosterDragon replied:
Ah you are correct. Fixed.

PunkPun commented:
```
The abstract path should never be searched for an unreachable point. Abstract cell 154,34 failed to route to abstract cell.
```


Another crash
[crash2.orarep.zip](https://github.com/OpenRA/OpenRA/files/10157308/crash2.orarep.zip)

PunkPun commented:
Confirmed that this fixes the crash reported in https://forum.openra.net/viewtopic.php?p=314473#p314473

anvilvapre commented:
I've read the source for the first time. Some open questions. 

Wouldn't the following optimizations be possible?
1. If a grid (info) rebuild was triggered by actor blocking. Couldn't you then skip the `locomotor.MovementCostForCell()` call in `CellIsAccessible`? Or could the movement of an actor impact movement cost of surrounding cells? Al you would need to do is mark a grid to have become dirty because of actor map changes - and not cell cost changes? 

2. It seems very expensive to have to retrieve all edges of all neighbouring grids (infos) - each time when a grid (info) is rebuilt after it was marked dirty. You traverse the edges (cell on two sides of it) for 5 grids - when only one grid is updated? If a neighbouring grid also was marked dirty - you again build a list of edges and traverse most of the same edges there?

I wonder how this performs late in game with many/most grids being marked dirty each logic tick.

2.a. Is there no option to cache the edges? Only update the edges on the side of which you know they may have become invalid? I.e. if grid A is dirty. And if you query the edges of the grid left of A named B. Wouldn't it then mean that only the right side edges of grid B could ever be different?

2.b. Edge cells are - need to be - queried on a  local cell basis, not abstract cell basis. This means you still can't take advantage of one abstract cell representing a whole grid. If one grid can be represented by a single abstract cell - you stil check 4 sides times 10 cells times 3 adjecent cells. 

2.c. You query all edges of all adjecent grids of a grid A. Couldn't you limit it to only querying the edges that are connected to grid A? 


2.d Can't the order in which you rebuild dirty grids impact the abstract graph? I.e. if in a horizontal row of grids first the grid in column 0 is updated, and only later the grid of 1. Couldn't the edges returned for for grid in column 1 during the updating of grid in column 0 no longer be valid - after grid 1 was updated?

I.e. you have grid A, B, C in a horizontal row. A is marked dirty. B is marked dirty. B was represented by a single abstract cell before it was marked dirty. after it was marked dirty the whole grid was covered by many actors resulting in the grid needing to be represented by i.e. 10 abstract cells - where the number of edges with A were reduced. Wouldn't while rebuiling grid A - the old edges of B be traversed?

3. Perhaps good to throw an exception if the actor map is updated and/or the locomotor indicates the cell cost of a cell changed while a path search is active - i.e. a grid being marked dirty while a search was active. This in theory should not be possible - as everything runs in a single thread. However it might allow you to spot hard to find bug where a search causes cell cost to change and/or the actor map to change.

RoosterDragon commented:
> ```
> The abstract path should never be searched for an unreachable point. Abstract cell 154,34 failed to route to abstract cell.
> ```
> 
> Another crash [crash2.orarep.zip](https://github.com/OpenRA/OpenRA/files/10157308/crash2.orarep.zip)

Have pushed a change which prevents the crash on this replay plus I reverified all the other replays remain resolved. The change is teaches the path search that a `heuristic` is allowed to provide a cost that indiciates a cell in unreachable - preventing it being explored at all.

Previously, we provided a high cost. This meant if a path could be found in the intended direction we would avoid exploring the unreachable area. However if the intended direction was also blocked (in the crash2 replay due to another sub in the way) then we returned to searching in the bad area and caused a crash. With this change we don't search in the bad area at all.

----

@anvilvapre a separate issue for your comment would be great so it doesn't get lost once this PR closes.

dnqbob commented:
going to test if https://github.com/OpenRA/OpenRA/pull/20498/commits/52e8fb37bfaec7b599adacba3e96b31b42d0efbc fix the related issue #20528

update: nope

PunkPun commented:
While playing around I couldn't catch any more crashes. Though #20528 is reported to still not be fixed. I can't double check as my PC is incompatible with their branch

RoosterDragon committed:
HPF is aware of map projection changes.

An event is added to Map to indicate when the cell projection is changed. This is important as this can mean Map.Contains(CPos) could now return different results for the cell. The HierarchicalPathFinder is made aware of these changes so it can rebuild any out-of-date information. This fixes prevent a crash if a cell that was previously outside the map changes height and becomes inside the map. The local path search will explore the cell as it is inside the map - but if the HPF was unaware if had been updated, it will still consider the cell to be outside the map and unreachable, resulting in a crash.

RoosterDragon commented:
I have pushed an additional commit that fixes #20528 - although a bit of scope creep on the original goal it's all related to fixing HPF crashes so seemed reasonable enough to bring along into this PR.

----

HPF is aware of map projection changes.

An event is added to Map to indicate when the cell projection is changed. This is important as this can mean Map.Contains(CPos) could now return different results for the cell. The HierarchicalPathFinder is made aware of these changes so it can rebuild any out-of-date information. This fixes prevent a crash if a cell that was previously outside the map changes height and becomes inside the map. The local path search will explore the cell as it is inside the map - but if the HPF was unaware if had been updated, it will still consider the cell to be outside the map and unreachable, resulting in a crash.

anvilvapre commented:
when can this happen? in the editor? or also in game?

PunkPun commented:
It can't happen in the editor as pathfinder isn't present there

pchote reviewed (approved):
Looks sensible, and the motivation makes sense. Lets take this and see what wider testing reveals.

pchote commented:
2392566c1d578c72fb065e56dc1485e31f5fa12f
d6d77eab7cdb21545aa4f772aa51d53c5287e66a

https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/e61f4427298111ba4d87bbd448fde6d9b545109b

dnqbob commented:
> when can this happen? in the editor? or also in game?

I think it may due to our branch (OpenMeow) change of:
![build](https://user-images.githubusercontent.com/13763394/208565689-59bc8c14-97ed-4f3a-af32-46100f88b00d.gif)

We make terrain height can be changed, and allow buildings can be built and change the terrain height, but we don't think this will change the pathfinding related (terrain type won't change and cliff will not affected by this function) and following is very strange:

> This fixes prevent a crash if a cell that was previously outside the map changes height and becomes inside the map.

dnqbob commented:
The motivation we do this function:

(1). In westwood's TS, terrain height can change as some weapons can change the terrain height. We already makes a warhead that can do that, although not applied in game for now.

(2). (1) leads to a problem in westwood TS: the player cannot build building on the bombarded place, which heavily affect the late gameplay (you can hardly to find a place to build a building or deploy a MCV), so we make the buildings can be built on rough and change the terrain height if the terrain type beneath is valid. The similar method is also used by the 3rd generation game of cnc, for example, General.

In conclusion, we want to make the terrain height change mechanic back from TS while making it player friendly as well.

dnqbob commented:
> This fixes prevent a crash if a cell that was previously outside the map changes height and becomes inside the map.		

This is our function of change terrain height:

```
public void FlatCellWithHeight(World world, CPos cell, int height)
		{
			if (!CellInfos.Contains(cell))
				return;

			CellInfos[cell].FlatCell(height, this);

			UpdateCellVertex(world, cell);
		}
```
I have a question: doesn't `if (!CellInfos.Contains(cell))` skip the cell outside the map?

RoosterDragon commented:
In the replay from your issue here is the cell (39,-28) in question. We can see it is outside the map when the game starts. I am using the "All Players" shroud view and the `/terrain-geometry` debug command.
![image](https://user-images.githubusercontent.com/3399086/208687913-cde7b338-e85c-4a4b-b479-83f083549712.png)

Once the power plant is placed at 10:58, the land flattening nearby brings the cell onto the map.
![image](https://user-images.githubusercontent.com/3399086/208689562-bf502d3f-d087-4663-8373-2601fcbdc8a2.png)

The path search at 11:07 was then crashing because the local path search was trying to explore the cell, which is valid as it is on the map, but HPF still thought that cell was not on the map and crashed. The fix applied means HPF will now be told about the change and will know the cell is now inside the map.

So it looks like although the cell wasn't directly flattened, but flattening the nearby land caused it to become level enough to count.

dnqbob commented:
Thanks, that is very important for us! Going to do some reproduction on this and look into this issue.

dnqbob commented:
Another question: will https://github.com/OpenRA/OpenRA/commit/d6d77eab7cdb21545aa4f772aa51d53c5287e66a be expensive when updating the projection cells and pathfinding cache? Considering we will have weapon that can change terrain height like:
![terrain](https://user-images.githubusercontent.com/13763394/208893541-ef4a7999-9b05-4872-979c-1aa577ccc13f.gif)

RoosterDragon commented:
I wouldn't expect you to see problems from that weapon alone, because it's likely to cause only a couple of the 10x10 grids that HPF uses to need to be refreshed. Dropping lots of bombs in a small area isn't likely to be a problem.

If you dropped bombs over a very wide area of the map all at once, you would cause more grids to need to be refreshed and might be able to see a lag - but unless you actually run into such an issue I wouldn't worry about it.

dnqbob commented:
Thanks!
